src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs

```diff
@@ -142,30 +181,153 @@ public async Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object> arg
+                using var stream = new ServerDirectStream();
+
+                Task? task = null;
+                try
+                {
+                    // send request to asset source
+                    task = _rpc.InvokeWithCancellationAsync(targetName, arguments.Concat(stream.Name).ToArray(), cancellationToken);
+
+                    // if invoke throws an exception, make sure we raise cancellation.
+                    RaiseCancellationIfInvokeFailed(task, mergedCancellation, cancellationToken);
+
+                    // wait for asset source to respond
+                    await stream.WaitForDirectConnectionAsync(mergedCancellation.Token).ConfigureAwait(false);
+
+                    // run user task with direct stream
+                    var result = await funcWithDirectStreamAsync(stream, mergedCancellation.Token).ConfigureAwait(false);
+
+                    // wait task to finish
+                    await task.ConfigureAwait(false);
+
+                    return result;
+                }
+                catch (Exception ex) when (ReportUnlessCanceled(ex, mergedCancellation.Token, cancellationToken))
+                {
+                    // important to use cancelationToken here rather than mergedCancellationToken.
+                    // there is a slight delay when merged cancellation token will be notified once cancellation token
+                    // is raised, it can cause one to be in cancelled mode and the other is not. here, one we
+                    // actually care is the cancellation token given in, not the merged cancellation token.
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    // record reason why task got aborted. use NFW here since we don't want to
+                    // crash VS on explicitly killing OOP.
+                    (task?.Exception ?? ex).ReportServiceHubNFW("JsonRpc Invoke Failed");
+
+                    throw;
+                }
             }
-            catch (Exception ex) // no when since Extensions.InvokeAsync already recorded it
+            catch (Exception ex)
             {
                 throw CreateSoftCrashException(ex, cancellationToken);
             }
         }
 
-        public Task<TResult> InvokeAsync<TResult>(
+        public async Task<TResult> InvokeAsync<TResult>(
             string targetName, IReadOnlyList<object> arguments, Func<Stream, CancellationToken, TResult> funcWithDirectStream, CancellationToken cancellationToken)
         {
             Contract.ThrowIfFalse(_startedListening);
             cancellationToken.ThrowIfCancellationRequested();
 
             try
             {
-                return Extensions.InvokeAsync(_rpc, targetName, arguments, funcWithDirectStream, cancellationToken);
+                using var mergedCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```

Comment:
i dislike this duplication because there's no way someone would know above that i fthey fixed something in InvokeAsync that they would also have to fix the same thing here.  We should unify the code as much as we can, and/or also just comment that these methods must be kept in sync.

Reply:
It's removed. Check the latest version.